xiilib/databases.py

```diff
@@ -54,16 +56,16 @@ def __init__(self, charm: ops.CharmBase, relation_name: str):
             charm: The requesting charm object.
             relation_name: The redis relation name.
         """
-        self._charm = charm
-        self.relation_name = relation_name
-        # redis charm doesn't provide database id via the relation
+        # we don't need store for Redis
+        self._redis = RedisRequires(charm=charm, _stored=MagicMock(), relation_name=relation_name)
         self.database = ""
-        self.on = self._RedisDatabaseRequiresEventShim(
-            self._charm.on[relation_name].relation_changed
-        )
+        self.relation_name = relation_name
+        relation = charm.model.get_relation(relation_name)
+        self.relation_id = relation.id if relation else None
+        self.on = self._RedisDatabaseRequiresEventShim(charm.on.redis_relation_updated)
 
-    def fetch_relation_data(
-        self, fields: typing.List[str]
+    def fetch_relation_data(  # argument fields is used for matching the function signature
```

Comment:
What function signature is being matched here please?

Reply:
It's for matching `charms.data_platform_libs.v0.data_interfaces.DataRelation.fetch_relation_data`.

Reply:
What is the reason for needing to match it please?

Reply:
Well, the shim needs to match the function it replaced; that's the definition of a shim.